File: src/dsl/parser.rs
use crate::dsl::{Expr, Token, Lexer, BinaryOp, UnaryOp};

#[derive(Debug)]
pub enum Stmt {
    Let(String, Expr),
    Fun(String, Vec<String>, Expr),
    Expr(Expr),
}

#[derive(Debug)]
pub struct Parser{
    tokens: Vec<Token>,
    pos: usize,
}

impl Parser{
    pub fn new(input: &str) -> Result<Self, String>{
        let mut lexer = Lexer::new(input);
        Ok(Parser{
            tokens: lexer.tokenize()?,
            pos: 0,
        })
    }

    pub fn parse(&mut self) -> Result<Stmt, String> {
        if let Some(token) = self.peek() {
            match token {
                Token::Let => {
                    self.next();
                    self.parse_let()
                }
                Token::Fun => {
                    self.next();
                    self.parse_fun()
                }
                _ => {
                    let expr = self.parse_expr()?;
                    Ok(Stmt::Expr(expr))
                }
            }
        }else{
            Err("unexpected error.".into())
        }
    }

    fn peek(&self) -> Option<&Token> {
        self.tokens.get(self.pos)
    }

    fn next(&mut self) -> Option<&Token> {
        if self.pos < self.tokens.len() {
            self.pos += 1;
        }
        self.tokens.get(self.pos - 1)
    }

    fn consume(&mut self, expected: &Token) -> bool {
        if let Some(token) = self.peek() {
            if token == expected {
                true
            }else{
                false
            }
        }else{
            false
        }
    }



    fn parse_let(&mut self) -> Result<Stmt, String> {
        let name = match self.next(){
            Some(Token::Ident(ident)) => ident.clone(),
            _ => {
                return Err("Expected identifier after 'let'".into());
            }
        };
        if !self.consume(&Token::Equal){
            return Err("Expected '=' in let binding".into());
        }

        let expr = self.parse_expr()?;
        Ok(Stmt::Let(name, expr))
    }

    fn parse_fun(&mut self) -> Result<Stmt, String> {
        let name = match self.next(){
            Some(Token::Ident(ident)) => ident.clone(),
            _ => {
                return Err("Expected identifier after 'fun'".into());
            }
        };
        if !self.consume(&Token::Equal) {
            return Err("Expected '=' in fun definition".into());
        }

        // (x, y, ...) -> body
        if !self.consume(&Token::LParen){
            return Err("Expected '(' after 'fun f ='".into());
        }
        let mut params = Vec::new();
        loop{
            match self.next(){
                Some(Token::Ident(id)) => params.push(id.clone()),
                Some(Token::RParen) => { break; },
                Some(Token::Comma) => { continue; },
                _ => {
                    return Err("Invalid parameter list in function definition".into());
                }
            }

            if self.consume(&Token::RParen){
                break;
            }else if self.consume(&Token::Comma){
                continue;
            }
        }

        if !self.consume(&Token::Arrow){
            return Err("Expected '->' in function definition".into());
        }

        let body = self.parse_expr()?;
        Ok(Stmt::Fun(name, params, body))
    }

    fn parse_expr(&mut self) -> Result<Expr, String> {
        self.parse_add_sub()
    }

    fn parse_add_sub(&mut self) -> Result<Expr, String> {
        let mut node = self.parse_mul_div()?;
        loop{
            match self.peek(){
                Some(Token::Plus) => {
                    self.next();
                    let rhs = self.parse_mul_div()?;
                    node = Expr::Binary(BinaryOp::Add, Box::new(node), Box::new(rhs));
                }
                Some(Token::Minus) => {
                    self.next();
                    let rhs = self.parse_mul_div()?;
                    node = Expr::Binary(BinaryOp::Sub, Box::new(node), Box::new(rhs));
                }
                _ => { break; }
            }
        }
        Ok(node)
    }

    fn parse_mul_div(&mut self) -> Result<Expr, String> {
        let mut node = self.parse_pow()?;
        loop {
            // implicit multiplication: if next token starts with a primary, treat as '*'
            let implicit_mul = matches!(
                self.peek(),
                Some(Token::Num(_)) | Some(Token::Ident(_)) | Some(Token::LParen)
            );
            match self.peek() {
                Some(Token::Star) => {
                    self.next();
                    let rhs = self.parse_pow()?;
                    node = Expr::Binary(BinaryOp::Mul, Box::new(node), Box::new(rhs));
                }
                Some(Token::Slash) => {
                    self.next();
                    let rhs = self.parse_pow()?;
                    node = Expr::Binary(BinaryOp::Div, Box::new(node), Box::new(rhs));
                }
                _ if implicit_mul => {
                    let rhs = self.parse_pow()?;
                    node = Expr::Binary(BinaryOp::Mul, Box::new(node), Box::new(rhs));
                }
                _ => { break; }
            }
        }
        Ok(node)
    }

    fn parse_pow(&mut self) -> Result<Expr, String> {
        let mut node = self.parse_unary()?;
        while let Some(Token::Caret) = self.peek() {
            self.next();
            let rhs = self.parse_unary()?;
            node = Expr::Binary(BinaryOp::Pow, Box::new(node), Box::new(rhs));
        }
        Ok(node)
    }

    fn parse_unary(&mut self) -> Result<Expr, String> {
        match self.peek(){
            Some(Token::Minus) => {
                self.next();
                let expr = self.parse_unary()?;
                Ok(Expr::Unary(UnaryOp::Neg, Box::new(expr)))
            }
            _ => self.parse_primary(),
        }
    }

    fn parse_primary(&mut self) -> Result<Expr, String> {
        todo!("")
    }
}

File: creola-core/parser.cpp
#include "parser.hpp"
#include<cctype>
#include<stdexcept>


// -*----------------------------------------------------------------*-
// -*- begin::namespace::creola                                     -*-
// -*----------------------------------------------------------------*-
namespace creola::lexer{
//
Tokenizer::Tokenizer(std::string&& src)
: m_src{std::move(src)}, m_pos{0}
{}

Token Tokenizer::next(void){
    this->skip_whitespace();
    if(this->m_pos == this->m_src.size()){
        return Token(TokenKind::End, "");
    }

    auto c = this->m_src[this->m_pos];
    if(std::isdigit(c)){
        auto start = this->m_pos;
        size_t& pos = this->m_pos;
        while(pos < this->m_src.size() && (std::isdigit(this->m_src[pos]) || this->m_src[pos]=='.')){
            ++pos;
        }
        auto lexme = this->m_src.substr(start, pos-start);
        auto num = std::stod(lexme);
        return Token(TokenKind::Number, lexme, num);
    }
    if(std::isalpha(c) || c == '_'){
        auto start = this->m_pos;
        size_t& pos = this->m_pos;
        while(pos < this->m_src.size() && (std::isalnum(this->m_src[pos]) || this->m_src[0]=='_')){
            ++pos;
        }
        auto lexeme = this->m_src.substr(start, pos-start);
        if(lexeme=="let"){
            return Token(TokenKind::KwLet, "let");
        }
        if(lexeme=="fun"){
            return Token(TokenKind::KwFun, "fun");
        }
        return Token(TokenKind::Ident, lexeme);
    }
    ++this->m_pos;
    switch(c){
    case '+': return Token(TokenKind::Plus, "+");
    case '-': return Token(TokenKind::Minus, "-");
    case '*': return Token(TokenKind::Star, "*");
    case '/': return Token(TokenKind::Slash, "/");
    case '^': return Token(TokenKind::Caret, "^");
    case '(': return Token(TokenKind::LParen, "(");
    case ')': return Token(TokenKind::RParen, ")");
    case ',': return Token(TokenKind::Comma, ",");
    case '=': return Token(TokenKind::Equal, "=");
    }

    return Token(TokenKind::End, "");
}

void Tokenizer::skip_whitespace(void){
    while(this->m_pos < this->m_src.size()){
        auto c = this->m_src[this->m_pos++];
        if(std::isspace(c)){ break; }
    }
}

// ---
Parser::Parser(std::string&& src)
: m_tokenizer{std::move(src)}{
    this->m_curTok = this->m_tokenizer.next();
}

core::Expr Parser::parse(void){
    //! @todo
}


void Parser::consume(TokenKind kind){
    if(this->match(kind)){
        this->m_curTok = this->m_tokenizer.next();
    }
}

void Parser::expect(TokenKind kind){
    if(!this->match(kind)){
        throw std::runtime_error("unexpected token");
    }
}


core::Expr Parser::parse_primary(void){
    if(this->match(TokenKind::Number)){
        auto val = this->m_curTok.num;
        this->consume(TokenKind::Number);
        return core::number(val);
    }
    if(this->match(TokenKind::Ident)){
        auto name = this->m_curTok.text;
        this->consume(TokenKind::Ident);
        if(this->match(TokenKind::LParen)){
            this->consume(TokenKind::LParen);
            Vec<core::Expr> args{};
            if(!this->match(TokenKind::RParen)){
                while(true){
                    args.push_back(this->parse());
                    if(this->match(TokenKind::Comma)){
                        this->consume(TokenKind::Comma);
                    }else{
                        break;
                    }
                }
            }
            this->expect(TokenKind::RParen);
            this->consume(TokenKind::RParen);
            return std::make_shared<core::FuncCall>(name, args);
        }
        return core::symbol(name);
    }
    if(this->match(TokenKind::LParen)){
        this->consume(TokenKind::LParen);
        auto expr = this->parse();
        this->expect(TokenKind::RParen);
        this->consume(TokenKind::RParen);
        return std::move(expr);
    }
    if(this->match(TokenKind::Minus)){
        this->consume(TokenKind::Minus);
        return std::make_shared<core::Neg>(this->parse_primary());
    }

    throw std::runtime_error("invalid primary");
}

core::Expr Parser::parse_pow(void){
    auto lhs = this->parse_primary();
    while(this->match(TokenKind::Caret)){
        this->consume(TokenKind::Caret);
        auto rhs = this->parse_primary();
        lhs = std::make_shared<core::Pow>(lhs, rhs);
    }
}
core::Expr Parser::parse_term(void){
    //! @todo
}

/*
// -*-
class Parser{
public:
private:
Tokenizer m_tokenizer;
Token m_curTok;


};
*/

// -*----------------------------------------------------------------*-
}//-*- end::namespace::creola                                       -*-
// -*----------------------------------------------------------------*-